Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp

```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
 	if (PlayerPawnBPClass.Class != NULL)
 	{
 		DefaultPawnClass = PlayerPawnBPClass.Class;
+		PlayerCheckoutRadius = FGenericPlatformMath::Sqrt(PlayerPawnBPClass.Class.GetDefaultObject()->NetCullDistanceSquared);
 	}
-	
+
+	PlayersSpawned = 0;
+
 	// Seamless Travel is not currently supported in SpatialOS [UNR-897]
 	bUseSeamlessTravel = false;
 }
+
+void ABigGymGameMode::InitGameState()
+{
+	if (ShouldUseCustomSpawning())
+	{
+		ParsePassedValues();
+		ClearExistingSpawnPoints();
+		GenerateSpawnPoints(TotalPlayers);
+	}
+}
+
+bool ABigGymGameMode::ShouldUseCustomSpawning()
+{
+	FString WorkerValue;
+	USpatialWorkerFlags::GetWorkerFlag(TEXT("override_spawning"), WorkerValue);
+	return (WorkerValue.Equals(TEXT("true"), ESearchCase::IgnoreCase) || FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")));
+}
+
+void ABigGymGameMode::ParsePassedValues()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")))
+	{
+		if (FParse::Param(FCommandLine::Get(), TEXT("TotalPlayers=")))
+		{
+			FParse::Value(FCommandLine::Get(), TEXT("TotalPlayers="), TotalPlayers);
+		}
+	}
+	else
+	{
+		FString TotalPlayersString;
+
+		if (USpatialWorkerFlags::GetWorkerFlag(TEXT("total_players"), TotalPlayersString))
+		{
+			TotalPlayers = FCString::Atoi(*TotalPlayersString);
+		}
+	}
+}
+
+void ABigGymGameMode::ClearExistingSpawnPoints()
+{
+	TArray<AActor*> ExistingSpawnPoints;
+
+	UGameplayStatics::GetAllActorsOfClass(GetWorld(), APlayerStart::StaticClass(), ExistingSpawnPoints);
+
+	for (AActor* SpawnPoint : ExistingSpawnPoints)
+	{
+		SpawnPoint->Destroy();
+	}
+}
+
+void ABigGymGameMode::GenerateSpawnPoints(int SpawnPointsNum)
+{
+	// Spawn in the air above terrain obstacles.
+	int Z = 305;
+
+	if (SpawnPointsNum < 1)
+	{
+		UE_LOG(LogTemp, Error, TEXT("Can't spawn requested %d spawn points - the spawn points number should be at least 1"), SpawnPointsNum);
+		return;
+	}
+
+	int NumRows = FMath::CeilToInt(FGenericPlatformMath::Sqrt(SpawnPointsNum));
+	int NumCols = FMath::CeilToInt(SpawnPointsNum / (float)NumRows);
+
+	int DistBetweenSpawnPoints = 300;
```

Comment:
This seems to be in SpatialOS units, whereas in Unreal 100 units = 1 SpatialOS unit?

Reply:
Discussed offline - it's intentional and it's so that players spawn in a grid close to each other.

Reply:
Worth clarifying if this is supposed to be meters (spatial) or centimetres (unreal)

Reply:
Worth clarifying if this is supposed to be meters (spatial) or centimetres (unreal)

Reply:
Done